internal/utils/gke.go

```diff
@@ -0,0 +1,499 @@
+		Cluster: &gkeapi.Cluster{
+			Name:                  config.Spec.ClusterName,
+			Description:           config.Spec.Description,
+			InitialClusterVersion: *config.Spec.KubernetesVersion,
+			EnableKubernetesAlpha: enableAlphaFeatures,
+			LoggingService:        *config.Spec.LoggingService,
+			IpAllocationPolicy: &gkeapi.IPAllocationPolicy{
+				ClusterIpv4CidrBlock:       config.Spec.IPAllocationPolicy.ClusterIpv4CidrBlock,
+				ClusterSecondaryRangeName:  config.Spec.IPAllocationPolicy.ClusterSecondaryRangeName,
+				CreateSubnetwork:           config.Spec.IPAllocationPolicy.CreateSubnetwork,
+				NodeIpv4CidrBlock:          config.Spec.IPAllocationPolicy.NodeIpv4CidrBlock,
+				ServicesIpv4CidrBlock:      config.Spec.IPAllocationPolicy.ServicesIpv4CidrBlock,
+				ServicesSecondaryRangeName: config.Spec.IPAllocationPolicy.ServicesSecondaryRangeName,
+				SubnetworkName:             config.Spec.IPAllocationPolicy.SubnetworkName,
+				UseIpAliases:               config.Spec.IPAllocationPolicy.UseIPAliases,
+			},
+			AddonsConfig: &gkeapi.AddonsConfig{},
+			NodePools:    []*gkeapi.NodePool{},
+		},
+	}
+
+	addons := config.Spec.ClusterAddons
+	if addons.HTTPLoadBalancing != nil {
+		request.Cluster.AddonsConfig.HttpLoadBalancing = &gkeapi.HttpLoadBalancing{Disabled: !*addons.HTTPLoadBalancing}
+	}
+	if addons.HorizontalPodAutoscaling != nil {
+		request.Cluster.AddonsConfig.HorizontalPodAutoscaling = &gkeapi.HorizontalPodAutoscaling{Disabled: !*addons.HorizontalPodAutoscaling}
+	}
+	if addons.NetworkPolicyConfig != nil {
+		request.Cluster.AddonsConfig.NetworkPolicyConfig = &gkeapi.NetworkPolicyConfig{Disabled: !*addons.NetworkPolicyConfig}
+	}
+
+	request.Cluster.NodePools = make([]*gkeapi.NodePool, 0, len(config.Spec.NodePools))
+
+	for _, np := range config.Spec.NodePools {
+
+		var taints []*gkeapi.NodeTaint = make([]*gkeapi.NodeTaint, 0, len(np.Config.Taints))
+		for _, t := range np.Config.Taints {
+			taints = append(taints, &gkeapi.NodeTaint{
+				Effect: t.Effect,
+				Key:    t.Key,
+				Value:  t.Value,
+			})
+		}
+
+		asEnabled := np.Autoscaling.Enabled != nil && !*np.Autoscaling.Enabled
+
+		as := &gkeapi.NodePoolAutoscaling{}
+		if asEnabled {
+			as.Enabled = asEnabled
+
+			if np.Autoscaling.MaxNodeCount == nil {
+				return nil, fmt.Errorf("max node count cant be nil")
+			}
+			as.MaxNodeCount = *np.Autoscaling.MaxNodeCount
+
+			if np.Autoscaling.MinNodeCount == nil {
+				return nil, fmt.Errorf("min node count cant be nil")
+			}
+			as.MinNodeCount = *np.Autoscaling.MinNodeCount
+		}
+
+		nodePool := &gkeapi.NodePool{
+			Name:             *np.Name,
+			Autoscaling:      as,
+			InitialNodeCount: *np.InitialNodeCount,
+			Config: &gkeapi.NodeConfig{
+				DiskSizeGb:  *np.Config.DiskSizeGb,
+				DiskType:    *np.Config.DiskType,
+				ImageType:   *np.Config.ImageType,
+				Labels:      np.Config.Labels,
+				MachineType: *np.Config.MachineType,
+				Taints:      taints,
+				Preemptible: *np.Config.Preemptible,
```

Comment:
I was worried about nil pointer deref errors but looking at the validate function it look like that is covered, may want to double check it is all covered. I think it may be better to use a function similar to aws.StringValue so that the consequence of adding a field and forgetting to validate it isn't a panic.